framework/aster-frame/src/mm/page_table/frame.rs

```diff
@@ -0,0 +1,515 @@
+//!    - or a handle to a page table frame,
+//!    - or a processor.
+//! This is implemented by using a reference counter in the frame metadata. If the above
+//! conditions are not met, the page table frame is ensured to be freed upon dropping the last
+//! reference.
+//!
+//! One can acquire exclusive access to a page table frame using merely the physical address of
+//! the page table frame. This is implemented by a lock in the frame metadata. Here the
+//! exclusiveness is only ensured for kernel code, and the processor's MMU is able to access the
+//! page table frame while a lock is held. So the modification to the PTEs should be done after
+//! the initialization of the entity that the PTE points to. This is taken care in this module.
+//!
+
+use core::{marker::PhantomData, mem::ManuallyDrop, ops::Range, panic, sync::atomic::Ordering};
+
+use super::{nr_subpage_per_huge, page_size, PageTableEntryTrait};
+use crate::{
+    arch::mm::{PageTableEntry, PagingConsts},
+    mm::{
+        paddr_to_vaddr,
+        page::{
+            allocator::FRAME_ALLOCATOR,
+            meta::{FrameMeta, PageMeta, PageTablePageMeta, PageUsage},
+            Page,
+        },
+        page_prop::PageProperty,
+        Frame, Paddr, PagingConstsTrait, PagingLevel, PAGE_SIZE,
+    },
+};
+
+/// The raw handle to a page table frame.
+///
+/// This handle is a referencer of a page table frame. Thus creating and dropping it will affect
+/// the reference count of the page table frame. If dropped the raw handle as the last reference,
+/// the page table frame and subsequent children will be freed.
+///
+/// Only the CPU or a PTE can access a page table frame using a raw handle. To access the page
+/// table frame from the kernel code, use the handle [`PageTableNode`].
+#[derive(Debug)]
+pub(super) struct RawPageTableNode<E: PageTableEntryTrait, C: PagingConstsTrait>
+where
+    [(); C::NR_LEVELS as usize]:,
+{
+    pub(super) raw: Paddr,
+    pub(super) level: PagingLevel,
+    _phantom: PhantomData<(E, C)>,
+}
+
+impl<E: PageTableEntryTrait, C: PagingConstsTrait> RawPageTableNode<E, C>
+where
+    [(); C::NR_LEVELS as usize]:,
+{
+    pub(super) fn paddr(&self) -> Paddr {
+        self.raw
+    }
+
+    /// Convert a raw handle to an accessible handle by pertaining the lock.
+    pub(super) fn lock(self) -> PageTableNode<E, C> {
+        let page = unsafe { Page::<PageTablePageMeta<E, C>>::restore(self.paddr()) };
+        debug_assert!(page.meta().level == self.level);
+        // Acquire the lock.
+        while page
+            .meta()
+            .lock
+            .compare_exchange(0, 1, Ordering::Acquire, Ordering::Relaxed)
+            .is_err()
+        {
+            core::hint::spin_loop();
+        }
+        // Prevent dropping the handle.
+        let _ = ManuallyDrop::new(self);
+        PageTableNode::<E, C> { page }
+    }
+
+    /// Create a copy of the handle.
+    pub(super) fn copy_handle(&self) -> Self {
```

Comment:
Why is adding a `copy_handle` method more preferable than implementing the `Clone` trait?

And yes. I am skeptical about all code locations where the concept of "handle" is mentioned.

Reply:
A detailed response is provided in following discussions.

Reply:
The revised version has used the term handle consistently. So I am ok with bringing up "handle". But putting "handle" aside, "copy" is not the most appropriate word as Rust has established the semantics difference of copy and clone. When the word copy is used, people assume that you are copying some raw bytes. But apparently, a handle is not some raw bytes. `Frame` does not implement `Copy`.